sopt/src/main/scala/dagr/sopt/util/Configuration.scala

```diff
@@ -0,0 +1,248 @@
+  */
+object Configuration extends ConfigurationLike {
+
+  // A sorted set tracking all the configuration keys that are requested
+  private[util] val RequestedKeys = collection.mutable.TreeSet[String]()
+
+  // The global configuration instance for dagr
+  protected var _config: Config = ConfigFactory.load()
+
+  /** Implement the abstract method from the [[Configuration]] trait to return the actual config instance. */
+  override protected def config = _config
+
+  object Keys {
+    val CommandLineName = "dagr.command-line-name"
+    val ColorStatus     = "dagr.color-status"
+    val SystemPath      = "dagr.path"
+  }
+
+  /**
+    * Initialize the configuration by loading configuration from the supplied path, and combining it with
+    * configuration information from the system properties (higher priority), application.conf and
+    * reference.conf files (lower priority).
+    */
+  def initialize(path: Option[Path]): Unit = path match {
+    case None    =>
+      this._config = ConfigFactory.load()
+    case Some(p) =>
+      // setAllowMissing(false) refers to allowing the file(!) to be missing, not values within the file
+      val options = ConfigParseOptions.defaults().setAllowMissing(false)
+      val localConfig = ConfigFactory.parseFile(p.toFile, options)
+
+      // This mimics the behaviour of ConfigFactory.load() but with the localConfig sandwiched in
+      this._config = ConfigFactory.defaultOverrides()
+        .withFallback(localConfig)
+        .withFallback(ConfigFactory.defaultApplication())
+        .withFallback(ConfigFactory.defaultReference())
+        .resolve()
+  }
+
+  /** Allows initialization with a custom configuration. */
+  protected def initialize(customConfig: Config): Unit = this._config = customConfig
+
+  /** Returns a sorted set of all keys that have been requested up to this point in time. */
+  private[util] def requestedKeys: SortedSet[String] = {
+    var keys = collection.immutable.TreeSet[String]()
+    keys ++= RequestedKeys
+    keys
+  }
+
+  def commandLineName: String = {
+    new Configuration { } configure(dagr.sopt.util.Configuration.Keys.CommandLineName, "Clp")
+  }
+}
+
+/**
+  * Trait that provides useful methods for resolving all kinds of things in configuration,
+  * that is then mixed into the Configuration object above, and used to create a concrete
+  * Configuration trait below.
+  */
+trait ConfigurationLike {
+  protected def config : Config
+
+  /** Date format for any error logging */
+  private val fmt = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss")
+
+  private val name = sanitizeSimpleClassName(getClass.getSimpleName)
+
+  /** Removes various characters from the simple class name, for scala class names. */
+  private def sanitizeSimpleClassName(className: String): String = {
+    className.replaceFirst("[$].*$", "")
+  }
+
+  /** The method to write any errors, by default writing to standard error. */
+  protected def error(parts: Any*) : Unit = {
```

Comment:
I was thinking more on this ... how about we remove this from this `Configuration` and in Dagr's configuration override `configure[T]()` to call super, trap the exception, log it, then rethrow it?